### PR TITLE
feat: Oracle provider service account masquerade

### LINKED ISSUE #2816: Allow custom service account for Oracle provider
**Is your feature request related to a problem? Please describe.**

Oracle provider can only use pod service account for authentication using workload identity. It should support any valid service account token.

**Describe the solution you'd like**

Add `serviceAccountRef *esmeta.ServiceAccountSelector` field to the Oracle provider. If this field is specified, use that service account instead of the pod service account for workload identity authentication.

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by anders-swanson**
## Problem Statement

Allow other service accounts to be used for workload identity via `oracle.serviceAccountRef`

## Related Issue

Fixes #2816 

## Proposed Changes

Added a `serviceAccountRef` field to the Oracle provider, which allows the user to specify a custom service account for authentication when using Workload Identity.

## Checklist

- [ ] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [ ] All commits are signed with `git commit --signoff`
- [ ] My changes have reasonable test coverage
- [ ] All tests pass with `make test`
- [ ] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

File: pkg/provider/oracle/oracle.go [hunk 1]
```diff
@@ -394,9 +378,45 @@ func (vms *VaultManagementService) ValidateStore(store esv1beta1.GenericStore) e
...
+
 	return nil
 }
 
+func (vms *VaultManagementService) getWorkloadIdentityProvider(serviceAcccountRef *esmeta.ServiceAccountSelector, region, namespace string) (common.ConfigurationProvider, error) {
+	defer func() {
+		_ = os.Unsetenv(auth.ResourcePrincipalVersionEnvVar)
+		_ = os.Unsetenv(auth.ResourcePrincipalRegionEnvVar)
+		vms.workloadIdentityMutex.Unlock()
+	}()
```
Comment (Skarlso) [inline pkg/provider/oracle/oracle.go:395]:
```suggestion
func (vms *VaultManagementService) getWorkloadIdentityProvider(serviceAcccountRef *esmeta.ServiceAccountSelector, region, namespace string) (_ common.ConfigurationProvider, err error) {
	defer func() {
		if uerr := os.Unsetenv(auth.ResourcePrincipalVersionEnvVar); uerr != nil {
			err = errors.Join(err, fmt.Errorf("unable to set OCI SDK environment variable %s: %w", auth.ResourcePrincipalRegionEnvVar, uerr))
		}

		if uerr := os.Unsetenv(auth.ResourcePrincipalRegionEnvVar); uerr != nil {
			err = errors.Join(err, fmt.Errorf("unabled to unset OCI SDK environment variable %s: %w", auth.ResourcePrincipalVersionEnvVar, uerr))
		}

		vms.workloadIdentityMutex.Unlock()
	}()
```

Let's not ignore those unsets? :)

Comment (anders-swanson) [reply pkg/provider/oracle/oracle.go:395]: Sounds good

**Comment by Skarlso** [comment]
This seems to be doing what is being required. :D But I lack the context knowledge yet to fully grasp if there might be missing something from anywhere else. :)

**Review by moolen — COMMENTED** [review]
Hey :wave: sorry for the delay, i took some time to have a look. Overall LGTM, just one security issue that should be fixed before merging :)

File: pkg/provider/oracle/serviceaccount.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+// NewTokenProvider creates a new TokenProvider for a given service account.
+func NewTokenProvider(clientset kubernetes.Interface, serviceAccountRef *esmeta.ServiceAccountSelector, namespace string) *TokenProvider {
+	// "api" is the default OCI workload identity audience.
+	audiences := []string{"api"}
+	if len(serviceAccountRef.Audiences) > 0 {
+		audiences = append(audiences, serviceAccountRef.Audiences...)
+	}
+	if serviceAccountRef.Namespace != nil {
+		namespace = *serviceAccountRef.Namespace
+	}
```
[the remark this answers is not in the document]
Comment (anders-swanson) [reply pkg/provider/oracle/serviceaccount.go:46]: Thanks, this makes sense. I pushed a little change to add the missing validation layer.

**Review by moolen — APPROVED** [review]
LGTM, thank you for adding workload identity, big enhancement :tada:

**Pull request merged**